### PR TITLE
--cutline option with tests

### LINKED ISSUE #62: Support cutlines
It would be useful to be able to pass a cutline to `rio mbtiles`. Ideally, something like:

```sh
$ rio mbtiles --cutline extents.geojson --zoom-levels 10..3 in.vrt out.mbtiles
```

### PR DESCRIPTION
**Pull request by sgillies**
Resolves #62

A cutline file contains a GeoJSON FeatureCollection, the feature geometries of which are polygons or multipolygons. A cutline affects mbtiles export in 3 ways:

1. the bounds of the exported mbtiles are restricted by the bounds of the cutline shapes.
2. only tiles intersecting with the cutline shapes are generated.
3. only pixels within the cutline shapes are exported into tiles.

### CODE REVIEW

File: mbtiles/scripts/cli.py [hunk 1]
```diff
@@ -189,6 +204,12 @@ def resolve_inout(
 @click.option(
     "--progress-bar", "-#", default=False, is_flag=True, help="Display progress bar."
 )
+@click.option(
+    "--cutline",
+    callback=extract_features,
+    default=None,
+    help="A GeoJSON feature collection to be used as a cutline. Only source pixels within the cutline features will be exported.",
```
Comment (sgillies) [inline mbtiles/scripts/cli.py:211]: Maybe this should say "Path to a GeoJSON FeatureCollection ..."?

File: mbtiles/scripts/cli.py [hunk 2]
```diff
@@ -285,6 +309,31 @@ def mbtiles(
...
+                    invtransform.b,
+                    invtransform.d,
+                    invtransform.e,
+                    invtransform.xoff,
+                    invtransform.yoff,
+                )
+                cutline_rev = shapely.affinity.affine_transform(
+                    cutline_src, shapely_matrix
+                )
+                warp_options["cutline"] = shapely.wkt.dumps(cutline_rev)
```
Comment (sgillies) [inline mbtiles/scripts/cli.py:335]: This is the second aspect of using a cutline: we use it as a warp cutline to fence pixels in/out.

File: mbtiles/scripts/cli.py [hunk 3]
```diff
@@ -285,6 +309,31 @@ def mbtiles(
...
             )
 
+            # cutlines must be transformed from CRS84 to src pixel/line
+            # coordinates and then formatted as WKT.
+            if cutline is not None:
+                geoms = [shape(f["geometry"]) for f in cutline]
+                union = unary_union(geoms)
+                if union.geom_type not in ("MultiPolygon", "Polygon"):
+                    raise click.ClickException("Unexpected cutline geometry type")
+                west, south, east, north = union.bounds
```
Comment (sgillies) [inline mbtiles/scripts/cli.py:319]: This is the first aspect of using a cutline: it restricts the bounds of the output mbtiles database.

File: mbtiles/scripts/cli.py [hunk 4]
```diff
@@ -348,10 +397,29 @@ def mbtiles(
...
+            # tiles.
+            if cutline:
+                geoms = [shape(f["geometry"]) for f in cutline]
+                union = unary_union(geoms)
+                cutline_mercator = transform_geom(
+                    "OGC:CRS84", "EPSG:3857", mapping(union)
                 )
+                min_ratio *= shape(cutline_mercator).area / raster_area
+                ratio = min_ratio
+                estimator = functools.partial(supermercado.burntiles.burn, cutline)
```
Comment (sgillies) [inline mbtiles/scripts/cli.py:411]: Here is the third aspect of the cutline: we use it to make an estimate of the number of output tiles.

File: mbtiles/scripts/cli.py [hunk 5]
```diff
@@ -466,9 +534,25 @@ def insert_results(tile, contents, img_ext=None, image_dump=None):
         def commit_mbtiles():
             conn.commit()
 
+        if cutline:
+
+            def gen_tiles():
+                for zk in range(minzoom, maxzoom + 1):
+                    for arr in supermercado.burntiles.burn(cutline, zk):
```
Comment (sgillies) [inline mbtiles/scripts/cli.py:541]: If we have a cutline, we use supermercado to find the intersecting tiles.

File: mbtiles/scripts/cli.py (hunk 1 above)
Comment (pratikyadav) [reply mbtiles/scripts/cli.py:211]: Yup sounds better.

**Review by pratikyadav — APPROVED** [review]
Looks good to me. The only suggestion would be to update readme.

**Pull request merged**